tests/test_pandas.py

```diff
@@ -246,6 +248,34 @@ def test_unique() -> None:
         ),
         np.ndarray,
     )
+    check(
+        assert_type(
+            pd.unique(pd.Index(["a", "b", "c", "a"])), Union[pd.Index, np.ndarray]
+        ),
+        np.ndarray,
```

Comment:
Interesting that it returns an `np.ndarray` here.  Docs say if input is an `Index`, it returns an `Index`. 

Aside from the special cases like `PeriodIndex`, are there any cases where it doesn't return `np.ndarray`?

Should try to figure that out, and maybe we can be more specific here, and then update the pandas docs.

Reply:
`unique` is pretty much a disaster of a function by design.  It should always do `Index[T]->Index[T]` but it doesn't.  For example, `Int64Index->ndarray`. I would also think if well designed it would always do `Series->Index` or `Series->ndarray` but not both. Series to Index would be the safest of the two.

Reply:
Also `DatetimeIndex -> DatetimeIndex`

Reply:
I could not do the full set of specific overload due to overlappign problems.  I tried all index types and being specific but failed due to inheritance and incompatible types.

Reply:
Maybe it isn't that bad.  Seems to be `ExtensionIndex -> ExtensionIndex` and otherwise `ndarray`

```python
 idx = [
pd.DatetimeIndex([pd.Timestamp(2000,1,1)]),
pd.Float64Index([1.0, 2.0]),
pd.Index(["a","b","c","a"]),
pd.Int64Index([1,2,3,4]),
pd.RangeIndex(0,10),
pd.UInt64Index([0,1,2,3,5]),
pd.timedelta_range("1 day","10 days",periods=10),
pd.MultiIndex.from_product([["a","b"],["apple","banana"]]),
pd.PeriodIndex(["2000Q1"],freq="Q"),
pd.CategoricalIndex(["a","b","a","c"]),
pd.interval_range(1,10,periods=10)
]

for val in idx:
    uniques = pd.unique(val)
    print(str(type(val)).split(".")[-1][:-2], "->", str(type(uniques)).split(".")[-1][:-2])
```

```plaintext
DatetimeIndex -> ndarray
Float64Index -> ndarray
Index -> ndarray
Int64Index -> ndarray
RangeIndex -> ndarray
UInt64Index -> ndarray
TimedeltaIndex -> ndarray
MultiIndex -> ndarray
PeriodIndex -> PeriodIndex
CategoricalIndex -> CategoricalIndex
IntervalIndex -> IntervalIndex

```